blog/2018/12/07/compound-selectors/index.markdown

```diff
@@ -0,0 +1,210 @@
+	<img class="robot" id="bender" src="...">
+	<img class="human" id="fry" src="...">
+	<img class="mutant" id="leela" src="...">
+
+With an argument I can choose any part that I like. Here I get the parts starting with the `li` tag:
+
+	$ perl html.pl li
+	<li><img class="robot" id="bender" src="..."></li>
+	<li><img class="human" id="fry" src="..."></li>
+	<li><img class="mutant" id="leela" src="..."></li>
+
+I can select all the images with a certain class:
+
+	$ perl html.pl img.human
+	<img class="human" id="farnworth " src="...">
+	<img class="human" id="fry" src="...">
+
+But what if I wanted just the human images in the list? I have to work a little bit harder. I specify a compound selector that notes that the `img` has to be in an `li` tag:
+
+	$ perl html.pl "li img.human"
+	<img class="human" id="fry" src="...">
+
+Imagine, then, more complicated HTML with other lists that also had images? I could add another selector to say it has to be in a certain sort of `ul` tag:
+
+	$ perl html.pl "ul.employees li img.human"
+	<img class="human" id="fry" src="...">
+
+If nothing should be between those tags. I can connect the selector with `>` to mean those should be immediate children instead of descendants:
+
+	$ perl html.pl "ul.employees > li > img.human"
+	<img class="human" id="fry" src="...">
+
+Now, consider how much work I've done there. Almost nothing. I made a DOM object, applied a selector, and I've isolated parts of the data. This is the same thing I was doing the hard way before. The think I like about Mojolicious though, is that this better way isn't more work (and I've built many things on top of [HTML::Parser](https://metacpan.org/pod/HTML::Parser)).
+
+## How about those new emojis?
+
+While writing about the [Unicode 9 updates in Perl v5.26](https://www.effectiveperlprogramming.com/2018/08/find-the-new-emojis-in-perls-unicode-support/), I wondered what I could show that might be interesting. How about figuring out which new emoji showed up?
+
+My first attempt simply trawled through every character and compared the various Unicode properties to see which code numbers changed from `Unassigned` to `Present_In`. That was find, but then I found that someone was already listing all the new emoji and I could scrape their site.
+
+I won't explain everything in this program. Trust me that it uses [Mojo::UserAgent](https://mojolicious.org/perldoc/Mojo/UserAgent) to fetch the data, extracts the DOM, and finds the text I want by using the compound selector `ul:not( [class] ) li a`. The rest is merely transforms on that extracted list. This is much easier than trying to do this with regexes:
+
+	use v5.28;
+	use utf8;
+	use strict;
+	use warnings;
+	use open qw(:std :utf8);
+	use charnames qw();
+
+	use Mojo::UserAgent;
+	my $ua = Mojo::UserAgent->new;
+
+	my $url = 'https://blog.emojipedia.org/new-unicode-9-emojis/';
+	my $tx = $ua->get( $url );
+
+	die "That didn't work!\n" unless $tx->success;
```

Comment:
`$tx->success` is deprecated. It was always just the inverse of `$tx->err` except that that also provides reasons. Suggest rewriting to `die "That didn't work!\n" if $tx->error;`.

Note that `$tx->result` dies on connection errors (which is what gives it value over the accessor `$tx->res` which is just the response message).